Translate micro-states to hydras

justbur opened the pull request:
They're basically the same thing anyway and this will allow us to have someone else maintain the backend code.

nixmaniack commented:
Cool! :+1: 
By the way, I did ask a question after looking at hydra on gitter. Was there any reason to not use hydra @syl20bnr?

scloudyy commented:
It's cool!

syl20bnr commented:
@nixmaniack no reason to not use them except a bad API based on colors at the time. We made the choice to use hydras coupled to some conventions like to not use color keywords and use explicit keywords instead. I'm not found of some unreadable docstrings in code so we can expect that I take measures to address them.

Also I would prefer that we use the term _micro-state_ instead of hydra. Hydra is the implementation and this term means nothing to users, _micro-state_ is the official term. This is equivalent to persps which are named _layouts_ in spacemacs.

syl20bnr commented:
I would have prefer hydras API to be more explicit like a `use-package` macro or even a `ivy-read` macro. But we should be able to keep things understandable with what I mentioned above.

tuhdo commented:
So, do the names got fixed now :D ?

syl20bnr commented:
Yes, we can use explicit keywords like `:foreign-keys` and `:exit` instead of a color that is a combination of both values. I don't like the term `:foreign-keys` to be honest but I'm fine with it, actually I don't like the term `hydra` too, something closer to what they really are like `define-transient-map` would have been better. Nothing stop us to alias the macro though. We can alias it `define-micro-state`.

syl20bnr commented on layers/+distribution/spacemacs-base/keybindings.el:
```diff
@@ -364,16 +364,14 @@
-  :disable-evil-leader t
-  :evil-leader "b."
-  :bindings
+(defhydra spacemacs/buffer-hydra (nil nil :hint nil)
```
Don't use hydra in the name, use `micro-state`.

syl20bnr replied:
Actually don't bother, I'll make a pass myself afterward.

robbyoconnor commented:
:eyes:

justbur commented:
I don't care about the name. I'd like it to not be too long, but that's all. Another option would be to create a new prefix for these things, whatever you want to call them. I actually just like the term state by itself. I think the micro part doesn't add much. "window-manipulation-state" indicates the idea to me just fine. I guess it's to avoid confusion with evil states, but they're all just modes (and mostly just keymaps) after all :wink:

On the unreadable docstring thing, this is a combination of two factors. The first is trying to create a table in a docstring, and the second is the markup used to color the keys. The underscores color the keys, but they don't take up horizontal space in the popup so there are characters that don't add horizontal space, "^", that allow you to align the tables in the docstring. 

We could not use the markup at all and color the string some other way. Anyway I didn't find them that bad after I understood why they look the way they do.

I don't like aliasing `defhydra`. @syl20bnr I know you don't like the name, but this makes the code less explicit for no real gain in my opinion. If you want people to understand the code you should use the actual function/macro that's being used. A spacemacs convention for hydras is a good idea. If the convention is complicated and you want to enforce it with a macro that wraps defhydra that's fine too. I don't see the benefit of an alias outweighing the downsides.

syl20bnr commented:
> I actually just like the term state by itself. I think the micro part doesn't add much. "window-manipulation-state" indicates the idea to me just fine. I guess it's to avoid confusion with evil states, but they're all just modes (and mostly just keymaps) after all 

I don't agree, micro-states are not the same things as states, they have different properties like the display of a docstring, an ephemeral concept with auto-leave on foreign keys etc... I'm for naming things clearly to avoid ambiguity while we discuss.

> Anyway I didn't find them that bad after I understood why they look the way they do.

I hope you are kidding me here, just look at the diffs of this PR if you are not convinced.

> If you want people to understand the code you should use the actual function/macro that's being used. 

I don't agree. If we use directly the hydra macros we won't be able to extent the concept of micro-states! I propose the following macro on top:

```elisp
(spacemacs|define-micro-state ..... :hydra <actual hydra>)
```

It allow us to perform Spacemacs specific stuff, we can imagine a registration of all hydras to be used as a source, or a check for `:color` keyword and display a warning if it is used with a message to guide the user toward the usage of `:foreign-keys` and `:exit`.

syl20bnr commented:
For the diffs of hydras docstring it is less bad on a big screen, on a 1680x1050 I get this...

![hydra_docstring](https://cloud.githubusercontent.com/assets/1243537/12326933/ad5a164e-baa1-11e5-95d1-5fc191f08287.png)

justbur commented:
> I don't agree, micro-states are not the same things as states, they have different properties like the display of a docstring, an ephemeral concept with auto-exit on foreign keys etc... and even their definition is different. I'm for naming things clearly to avoid ambiguity while we discuss.

Fine, I could argue with the micro terminology, since micro doesn't indicate ephemeral to me, but it's not worth it. micro-state is fine. 

> I hope you are kidding me here, just look at the diffs of this PR if you are not convinced.

It's just markup. It's *one* way to generate a docstring. Use whatever method you find most appealing.

syl20bnr commented:
> since micro doesn't indicate ephemeral to me

I have the same concern :-(

syl20bnr commented:
> It's just markup. It's one way to generate a docstring. Use whatever method you find most appealing.

I'm fine with it if we don't go wild, in this PR they are acceptable. I cannot find the example I saw somehwere about a very awful docstring full of `^`, this was insane.

As you suggested we could simplify the need for `_` in square brackets.

justbur commented:
I've gone through and adjusted all the macros, but I haven't tested extensively at all yet. Also, we need to figure out what to do about the docstrings. I fixed the easy ones, but there are a lot that could probably be made better.

StreakyCobra commented:
@justbur Amazing work, as usual. I tried a few of them and it works well. A cosmetic improvement would be to hide the cursor

justbur commented:
@StreakyCobra The odd thing is the cursor is supposed to be hidden already in that buffer. I don't see why it's showing up. It's hollow for me. Is that what you see?

StreakyCobra commented:
![window-manip-state](https://i.imgur.com/yOtMc6c.png)
![git-blame-state](https://i.imgur.com/sxQ5BqQ.png)

justbur commented:
Yea that's what I see too. Maybe it's a bug
On Fri, Jan 15, 2016 at 11:36 AM Fabien Dubosson <notifications@github.com>
wrote:

> [image: window-manip-state]
> <https://camo.githubusercontent.com/d638156e8e7014e24bc41275c873189b40b07709/68747470733a2f2f692e696d6775722e636f6d2f794f744d6336632e706e67>
> [image: git-blame-state]
> <https://camo.githubusercontent.com/a26bf5b7f28e6f201d0e472302ce425b4c338c7f/68747470733a2f2f692e696d6775722e636f6d2f737851354271512e706e67>
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/4606#issuecomment-172008998>.
>

StreakyCobra commented:
Let's say it's a feature then :joy:

justbur commented:
Nope it's a bug in evil. https://bitbucket.org/lyro/evil/pull-requests/28/evil-core-set-evil-local-mode-var-to-nil/diff

syl20bnr commented on core/core-micro-state.el:
```diff
@@ -265,4 +265,63 @@ pressed)."
+      the `:exit' keyword is evaluate *before* the actual execution
+      of the bound command.
+All properties supported by `spacemacs//create-key-binding-form' can be
+used."
```
Would be cool to support this as well see line 113. This is for quick binding of the micro-state and this is usable in both micro-states and toggles.

justbur replied:
ok, sure

justbur replied:
@syl20bnr added the key binding support

syl20bnr commented:
Okay.

syl20bnr commented:
I merged #4648 so this PR needs a rebase now.

StreakyCobra commented on layers/+source-control/version-control/funcs.el:
```diff
@@ -59,7 +59,7 @@
     (call-interactively
      (cl-case version-control-diff-tool
-       (diff-hl     'diff-hl-margin-minor-mode)
+       (diff-hl     'diff-hl-margin-mode)
```
Is it the only place where it need to be changed? Because the lines under looks pretty similar.

justbur replied:
Yes that function doesn't exist. Might just be a typo

StreakyCobra replied:
```text
diff-hl-margin-minor-mode is an interactive compiled Lisp function in
`diff-hl-margin.el'.
```

StreakyCobra replied:
And what about line 71 and 101 in the original file?

justbur replied:
Hm you're right. I was having a problem though. I can look at it again.

justbur replied:
@StreakyCobra This is the docstring for `diff-hl-margin-minor-mode`

  "Toggle displaying `diff-hl-mode' highlights on the margin locally.
You probably shouldn't use this function directly."

It's not autoloaded, which seems to be on purpose. Any idea why this function is used and not the plain mode?

StreakyCobra replied:
> It's not autoloaded, which seems to be on purpose. Any idea why this function is used and not the plain mode?

It has been introduced [here](https://github.com/syl20bnr/spacemacs/pull/3632/files#diff-7b025371ed8168e51c1dce9d7c919981R63).

@ralesi Did you use `diff-hl-margin-mode` in purpose, or would `diff-hl-mode` be fine? (ref. https://github.com/syl20bnr/spacemacs/issues/4034)

ralesi replied:
I haven't tested with the standard minor mode,  I think from other issues I have seen that we likely need to separate out fringe as a separate layer setting to make it more obvious which mode should be used. 


Rich






On Mon, Jan 18, 2016 at 6:59 AM -0800, "Fabien Dubosson" <notifications@github.com> wrote:












In layers/+source-control/version-control/funcs.el:
> @@ -59,7 +59,7 @@
>    (let ((current-prefix-arg t))
>      (call-interactively
>       (cl-case version-control-diff-tool
> -       (diff-hl     'diff-hl-margin-minor-mode)
> +       (diff-hl     'diff-hl-margin-mode)




It's not autoloaded, which seems to be on purpose. Any idea why this function is used and not the plain mode?




It has been introduced here.



@ralesi Did you use diff-hl-margin-mode in purpose, or would diff-hl-mode be fine? (ref. #4034)



—
Reply to this email directly or view it on GitHub.

ralesi replied:
This would also check if we are in the terminal or not. 


Rich






On Mon, Jan 18, 2016 at 6:59 AM -0800, "Fabien Dubosson" <notifications@github.com> wrote:












In layers/+source-control/version-control/funcs.el:
> @@ -59,7 +59,7 @@
>    (let ((current-prefix-arg t))
>      (call-interactively
>       (cl-case version-control-diff-tool
> -       (diff-hl     'diff-hl-margin-minor-mode)
> +       (diff-hl     'diff-hl-margin-mode)




It's not autoloaded, which seems to be on purpose. Any idea why this function is used and not the plain mode?




It has been introduced here.



@ralesi Did you use diff-hl-margin-mode in purpose, or would diff-hl-mode be fine? (ref. #4034)



—
Reply to this email directly or view it on GitHub.

justbur commented:
@syl20bnr rebased.

ralesi commented:
Any chance with the transition to hydra that we can add bindings to the hydra after its defined?  Thinking along the lines of something we can put in a `post-init`.

ralesi commented:
Also why don't we switch the naming to `transient-state` since noone seems to be completely happy with microstate or hydra?

nixmaniack commented:
> Any chance with the transition to hydra that we can add bindings to the hydra after its defined? Thinking along the lines of something we can put in a post-init.

Yes please! :+1:

justbur commented:
At first glance no hydras don't support this either, but I could probably put a hook in there to do this.

syl20bnr commented:
`transient-state` :purple_heart: 

I believe hydras support adding key bindings to it afterwards, I mean I believe that I read about it somewhere.

TheBB commented:
That would be great, we have half a dozen dirty hacks and just as many feature requests that would be solved by such a feature.

justbur commented:
hydras support inheritance, but I wouldn't want to use that. 

I have a way to do this in mind that should cover most cases.

justbur commented:
@ralesi @nixmaniack @TheBB here's a minimal example of a macro that adds bindings to a micro-state outside of its declaration. 
```elisp
(spacemacs|add-micro-state-bindings move
  ("f" forward-char "forward")
  ("b" backward-char "backward"))

(spacemacs|define-micro-state-2 move
  :doc
  "move"
  :bindings
  ("n" next-line "next")
  ("p" previous-line "previous"))

(global-set-key "\C-n" 'spacemacs/move-micro-state/next-line)
```

ralesi commented:
@justbur, this looks like it could be functional.  Two items I thought of:

- This looks like it only defines an additional variable, which then in turn is applied during execution of the `micro-state-2` macro.  How would we add mappings after execution of the macro (which I imagine will be performed on init?  Should we add a hook to re-execute the micro-state when the `add-to-micro-state-bindings` macro is run?
- Also since we are adding the bindings to a defvar, it might also be good idea to define a function to clear out the additional mappings if we want to start from scratch.

justbur commented:
@ralesi 

> This looks like it only defines an additional variable, which then in turn is applied during execution of the micro-state-2 macro. How would we add mappings after execution of the macro (which I imagine will be performed on init)? Should we add a hook to re-execute the micro-state when the add-to-micro-state-bindings macro is run?

You could I guess, but defining bindings in user-init will work fine. 

> Also since we are adding the bindings to a defvar, it might also be good idea to define a function to clear out the additional mappings if we want to start from scratch.

I don't see the use case for this one. We can make it more complicated if need be.

ralesi commented:
> You could I guess, but defining bindings in user-init will work fine.

What about the case of layer authors adjusting the micro-state?  Would the layer-init also run prior to micro-state execution?

justbur commented:
Can you give me an example? Why can't the layer redefine the micro state
On Tue, Jan 19, 2016 at 4:56 PM Rich Alesi <notifications@github.com> wrote:

> You could I guess, but defining bindings in user-init will work fine.
>
> What about the case of layer authors adjusting the micro-state? Would the
> layer-init also run prior to micro-state execution?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/syl20bnr/spacemacs/pull/4606#issuecomment-173000784>.
>

syl20bnr commented:
It should work in user-config, not user-init. The most probable use case is to add/modify key bindings of a transient state in a private layer that will be loaded at the end.

I did not watched the code but we could store the expanded form of the macro and a list of bindings so we can patch the expanded form at any time and evaluate it again. The downside is potential "namespace" pollution.

Another approach is to generate some kind of stubs to lazily evaluate the macro when the transient is first used. The downside is that modification to a transient state binding needs a restart.

ralesi commented:
> Can you give me an example? Why can't the layer redefine the micro state

One can definitely redefine the entire micro-state in a layer.  One use case along the lines of what I am suggesting is the VCS micro-state.  We should not have magit bindings in this micro-state unless the git / magit layer is also enabled.  The magit init should take care of adding this.

justbur commented:
Here's my opinion, but I can implement whatever you guys want. I think that there should be a simple convention for how bindings are added/changed. I don't see the problem with requiring that the new bindings be declared before the macro is initialized, since this is going to be a rare thing to do anyway and it will keep the code much simpler. 

If you only care about bindings, then it doesn't really matter if they are before or after, because you just need to know the keymap var. The only reason I can see to want to do it before defhydra is called is for the automatically generated docstring. The way it's designed now, you can add the bindings like in my example and have the docstring show them without changing the call to the macro. That's nice. if you don't care about the docstring they can be added at any time, no problem.

justbur commented:
Here's an update to the add/remove bindings thing. I think it would be simplest to have a convention like this. You can tell me if this is adequate. I'm just checking for a variable with the right name, which doesn't have to be defined. You can remove bindings by key or add bindings. You can also inspect the variables to see if you want to add or remove bindings. 

```elisp
(setq spacemacs-move-transient-state-remove-bindings '("p"))
(setq spacemacs-move-transient-state-add-bindings '(("f" forward-char "forward")))

(spacemacs|define-transient-state move
  :doc
  "move"
  :bindings
  ("n" next-line "next")
  ("p" previous-line "previous"))

(global-set-key "\C-n" 'spacemacs/move-transient-state/next-line)
```

syl20bnr commented:
The issue with using user-init is that we loose the possibility to update the transient-state at the layer level which is a big drawback.

But with one little trick we may be able to achieve the simplicity you mentioned while still being able to modify the transient states in layers: what if we defer the actual evaluation of all the transient-state macro calls to the very end of the initialization process ?

From an architectural point of view, being able to adjust transient state given the set of used layers is very interesting, the example of source control transient state is a good one.

justbur commented:
Yep we can delay execution. That makes sense to me and would be simple.

StreakyCobra commented:
And also for adapting the transient-state bindings in the `bepo` layer :innocent:

justbur commented:
@syl20bnr here's what I came up with

```elisp
(setq spacemacs-move-transient-state-remove-bindings '("p"))
(setq spacemacs-move-transient-state-add-bindings '(("f" forward-char "forward")))

(spacemacs|define-transient-state move
  :title "Move"
  :bindings
  ("n" next-line "next")
  ("p" previous-line "previous"))
(global-set-key "\C-n" 'spacemacs/move-transient-state/next-line)

(run-hooks 'spacemacs-transient-state-defs)
```

All of the declarations get loaded into `spacemacs-transient-state-defs` in a way that they can be run using `run-hooks` whenever you want. The bindings are calculated when the hook is run, but everything else is saved from the `spacemacs|define-transient-state` call.

syl20bnr commented:
Amazing PR :purple_heart: special mention for the new paste transient state, very neat, I think we can enable it by default now :+1: 
I reverted the git blame transient state, reason in the commit https://github.com/syl20bnr/spacemacs/commit/de44ab1
Cherry-picked into develop branch, you can safely delete your branch.

syl20bnr commented:
I had to do a little fix for eyebrowse due to TAB in docstring: https://github.com/syl20bnr/spacemacs/commit/7472991

bmag commented:
There's a bug with the layouts transient-state help text, please see #4822
